### PR TITLE
nixos/documentation: add option to generate caches

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change
An alternative approach to #76841 to generate manual pages caches. This will fix https://github.com/NixOS/nixpkgs/issues/14472.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change (mass-rebuild)
- [x] Tested execution of all binary files (`apropos`, `whatis`, `mandb` works as intended)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rnhmjoj** [comment]
Some notes:

- I tried to avoid using `symlinkJoin` by dumping all the paths in a `man.conf` inside the `runCommandLocal` and calling `mandb -C man.con`. However it kept overwriting the index for each path so that the index only contained the pages of the very last package.

- Maybe the patch should use a custom name instead of `man_db.conf` if we want to avoid indexing non-Nix packages on the system. I'm not sure whether this is expected or advisable.

**Review by matthewbauer — APPROVED** [review]

File: nixos/modules/misc/documentation.nix
Comment (Infinisil) [inline nixos/modules/misc/documentation.nix:220]: I don't think we need a man-db patch if we hardcode the cache path to e.g. `/etc/mandb`

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/misc/documentation.nix
Comment (rnhmjoj) [reply nixos/modules/misc/documentation.nix:220]: I'm not sure I understand. Are you proposing to symlink `manualCache` to  `/etc/mandb` or that the patch to check `/etc/man_db.conf` is not needed?

Comment (Infinisil) [reply nixos/modules/misc/documentation.nix:220]: We can compile it with a hardcoded config (which is supported without a patch) that adds a `MANDB_MAP` to `/etc/mandb`. Then we set `environment.etc.mandb.source = manualCache`

Comment (rnhmjoj) [reply nixos/modules/misc/documentation.nix:220]: Well, the patch is not great but it makes possible to configure man-db at "runtime", which may be useful besides adding the cache. Also `/etc` dosn't feel like the right place for a binary database.

File: nixos/modules/misc/documentation.nix
Comment (rnhmjoj) [reply nixos/modules/misc/documentation.nix]: Done, but shouldn't these function be documented? they are not in the nixpkgs manual.

Comment (rnhmjoj) [reply nixos/modules/misc/documentation.nix]: Done, but shouldn't these function be documented? they are not in the nixpkgs manual.

File: nixos/modules/misc/documentation.nix
Comment (lheckemann) [inline nixos/modules/misc/documentation.nix:202]: Maybe `buildEnv` would be more suitable, since it allows linking only `/share/man` of each package and requires fewer symlinks, e.g.

```
buildEnv {
  name = "man-paths";
  paths = config.environment.systemPackages;
  pathsToLink = ["/share/man"];
  extraOutputsToInstall = ["man"];
  ignoreCollisions = true;
}
```

File: nixos/modules/misc/documentation.nix
Comment (lheckemann) [reply nixos/modules/misc/documentation.nix:219]: IMHO it would be better to remove `"--with-config-file=${placeholder "out"}/etc/man_db.conf"` from the man-db expression and put `man_db.conf` into `/etc` with `environment.etc`.

**Review by lheckemann — COMMENTED** [review]
I like this!

File: nixos/modules/misc/documentation.nix
Comment (rnhmjoj) [reply nixos/modules/misc/documentation.nix:219]: @lheckemann That would be the best option but what about non-NixOS users?
If /etc/man_db.conf doesn't exist I don't think mandb programs will work at all (see https://github.com/NixOS/nixpkgs/issues/12097#issuecomment-168962194).
On debian there is no `man_db.conf`: the file is called `/etc/manpath.config`.

Comment (lheckemann) [reply nixos/modules/misc/documentation.nix:219]: Aah right, makes sense. Thanks for clearing that up :)

File: nixos/modules/misc/documentation.nix
Comment (rnhmjoj) [reply nixos/modules/misc/documentation.nix]: Done, but I haven't tested it yet because some major package appears to be broken in staging.

**Comment by eraserhd** [comment]
Bump, this looks nice and has been idle for 23 days.

**Comment by rnhmjoj** [comment]
@eraserhd: It's essentially ready. I'm waiting for @lheckemann and @infinisil to complete their review. If you want to try it out, cherry-pick the commits on master because staging is likely broken.

**Review by lheckemann — APPROVED** [review]

**Comment by lheckemann** [comment]
Sorry, didn't realise you were waiting for approval! *mind wanders to https://github.com/NixOS/rfcs/pull/30*

**Comment by rnhmjoj** [comment]
Documented the change in the 20.09 release notes.

**Comment by doronbehar** [comment]
Thank you @rnhmjoj for tackling this :heart:, the issue has bothered since always.

**Pull request merged**
